rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,574 @@
+  attestationStatement keys.AttestationStatement
+}
+
+// hardwareKeyAgentService has two implementations:
+//  - direct implementation with piv-go, adapted slightly from our existing implementation
+//  - hardware key agent gRPC service implementation
+type hardwareKeyAgentService interface {
+  GetInfo(ref hardwareKeyRef) hardwareKeyInfo
+  Sign(ref hardwareKeyRef, digest []byte, opts SignerOpts) (signature []byte, err error)
+}
+
+// Implements [crypto.Signer] and [keys.HardwareSigner].
+type hardwareKeyAgentKey struct {
+  agent  keyAgentService
+  keyRef hardwareKeyRef
+  // hardwareKeyInfo is cached after the first call
+  hardwareKeyInfo hardwareKeyInfo
+}
+
+// Implement [crypto.Signer]
+func (s *agentSigner) Public() crypto.PublicKey {
+  return s.hardwareKeyInfo.publicKey
+}
+
+// Implement [crypto.Signer]
+func (s *agentSigner) Sign(_ io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+  return s.agent.Sign(as.KeyRef, digest, opts)
+}
+
+// Implement [keys.HardwareSigner]
+func (s *agentSigner) GetAttestationStatement() *AttestationStatement {
+  return s.hardwareKeyInfo.attestationStatement
+}
+
+// Implement [keys.HardwareSigner]
+func (s *agentSigner) GetPrivateKeyPolicy() PrivateKeyPolicy {
+  return s.hardwareKeyInfo.privateKeyPolicy
+}
+
+func (s *agentSigner) fetchHardwareKeyInfo() (hardwareKeyInfo, error) {
+  if s.hardwareKeyInfo == nil {
+    s.hardwareKeyInfo = s.agent.GetInfo(s.ref)
+  }
+  return s.hardwareKeyInfo
+}
+```
+
+### UX
+
+#### PIN caching
+
+The PIN caching portion of this RFD does not present any additional UX concerns.
+It is purely a UX benefitting change, removing the need for back-to-back PIN
+prompts.
+
+#### Agent clients
+
+A user can start the hardware key agent with Teleport Connect or `tsh`.
+
+##### Teleport Connect
+
+Teleport Connect is an ideal runner for the key agent because it:
+
+1. Is a long-lived process
+1. Provides a UI for touch/PIN prompts, and can foreground itself for these prompts
+
+By default, if `tshd` detects a hardware key, it will automatically run and
+serve the hardware key agent service. This way, the agent will be served
+regardless of Teleport Connect's [login state](#running-the-agent-before-login).
+
+For example, if a user primarily wants to use `tsh`, but get PIN caching and
+PIN prompts in Teleport Connect, they could just launch Teleport Connect without
+logging in. Teleport Connect would just foreground itself with hardware key
+prompts for the user as needed without adding additional overhead.
+
+Alternative: By default, set `teleport.agent=false`. Once the user logs into
+Connect with a hardware key requirement for the first time, flip the flag to
+true indefinitely. The benefit with this approach is that we won't run an
+unused agent by default for users not using hardware key support. On the other
+hand, it requires the user to log in at least once for the feature to work as
+described in the example above.
```

Comment:
> The benefit with this approach is that we won't run an unused agent by default for users not using hardware key support.

What does it mean for an agent to be unused? A couple of lines above the RFD says this:

> By default, if `tshd` detects a hardware key, it will automatically run and serve the hardware key agent service.

Does it mean that any Yubikey plugged into the computer would make tshd in Connect start the agent?

Given what I said about reducing the exposure of the agent in the UI, I feel like we should try to avoid starting the agent unless we're certain that the user is going to need it.

Reply:
> Does it mean that any Yubikey plugged into the computer would make tshd in Connect start the agent?

Yes, it will always be run as there's always a chance that a `tsh` login session will require hardware key support. Otherwise it could require users to mirror every `tsh` login with a Teleport Connect login to get the agent running when it is needed.

Are you still concerned about this without the UI exposure of the feature?

Reply:
👍 

> I guess the cost goes up a little if we consider the hardware key agent as a possible attack vector.

This would be offset by the fact that we only serve Teleport keys in the first place - With no Teleport client keys in the hardware key, the hardware key agent would effectively provide nothing through the gRPC service.